qmlbind/src/metaobject.cpp

```diff
@@ -98,4 +106,165 @@ int MetaObject::metaCall(QObject *object, Call call, int index, void **argv) con
+
+        struct Placeholder {
+            MetadataBuilder *builder;
+            int index;
+            void setCurrentPos() {
+                builder->metadata[index] = builder->metadata.size();
+            }
+        };
+
+        MetadataBuilder &operator<<(uint value) {
+            metadata << value;
+            return *this;
+        }
+
+        Placeholder makePlaceholder() {
+            auto index = metadata.size();
+            metadata << 0;
+            return Placeholder { this, index };
+        }
+    };
+
+    StringDataBuilder strData;
+    MetadataBuilder metadata;
+
+    // content
+
+    metadata << 7; // revision
+    metadata << strData.add(className); // className
+    metadata << 0 << 0; // class info
+    metadata << mMethods.size(); // methods
+    auto methodOffsetPlaceholder = metadata.makePlaceholder();
+    metadata << mProperties.size(); // properties
+    auto propertyOffsetPlaceholder = metadata.makePlaceholder();
+    metadata << 0 << 0; // enums
+    metadata << 0 << 0; // constructors
+    metadata << 0; // flags
+
+    auto signalCount = std::count_if(mMethods.begin(), mMethods.end(), [](const MethodInfo &method) {
+        return method.type == MethodType::Signal;
+    });
+    metadata << signalCount; // signal count
+
+    methodOffsetPlaceholder.setCurrentPos();
+
+    // methods
+    QList<MetadataBuilder::Placeholder> paramPlaceholders;
+
+    for (const auto &method : mMethods) {
+        metadata << strData.add(method.name); // name
+        metadata << method.params.size(); // argc
+        paramPlaceholders << metadata.makePlaceholder(); // offset of parameters
+        metadata << 2; // tag
+        if (method.type == MethodType::Signal) {
+            metadata << 0x06; // flags
+        } else {
+            metadata << 0x02; // flags
+        }
+    }
+    auto jsValueType = 0x80000000 | strData.add("QJSValue");
+
+    // params
+    for (auto i = 0; i < mMethods.size(); ++i) {
+        paramPlaceholders[i].setCurrentPos();
+        const auto &method = mMethods[i];
+
+        // return type
+        if (method.type == MethodType::Signal) {
+            metadata << QMetaType::Void;
+        } else {
+            metadata << jsValueType;
+        }
+
+        // parameter types (all QJSValue)
+        for (auto i = 0; i < method.params.size(); ++i) {
+            metadata << jsValueType;
+        }
+
+        // parameter names
+        for (auto &&param : method.params) {
+            metadata << strData.add(param);
+        }
+    }
+
+    // properties
+    propertyOffsetPlaceholder.setCurrentPos();
+    for (const auto &prop : mProperties) {
+        metadata << strData.add(prop.name) << jsValueType;
+        auto signalIndex = mSignalIndexMap.value(prop.notifySignalName, -1);
+        if (signalIndex == -1) {
+            qWarning() << "qmlbind: no such notify signal" << prop.notifySignalName << "in class" << className;
+            metadata << 0x0009510b;
```

Comment:
You could add a comment here about what those two magic hex numbers in this if/else here mean.